sympy/stats/stochastic_process_types.py

```diff
@@ -566,6 +576,41 @@ def probability(self, condition, given_condition=None, evaluate=True, **kwargs):
                                 s -= self.probability(Eq(rv[0], i), Eq(rv[1], j)) * self.probability(Eq(rv[1], j), new_given_condition)
                     return s if greater else 1 - s
 
+            elif symbolic:
+                if isinstance(condition, Relational):
+
+                    curr_state = new_given_condition.rhs if isinstance(new_given_condition.lhs, RandomIndexedSymbol) \
+                            else new_given_condition.lhs
+                    next_state = condition.rhs if isinstance(condition.lhs, RandomIndexedSymbol) \
+                        else condition.lhs
+
+                    if isinstance(condition, Eq) or isinstance(condition, Ne):
+                        if isinstance(self, DiscreteMarkovChain):
+                            P = self.transition_probabilities**(rv[0].key - min_key_rv.key)
+                        else:
+                            P = exp(self.generator_matrix*(rv[0].key - min_key_rv.key))
+                        prob = P[curr_state, next_state] if isinstance(condition, Eq) else 1 - P[curr_state, next_state]
+                        return Piecewise((prob, rv[0].key > min_key_rv.key), (Probability(condition), True))
+                    else:
+                        upper = 1
+                        greater = False
+                        if isinstance(condition, Ge) or isinstance(condition, Lt):
+                            upper = 0
+                        if isinstance(condition, Gt) or isinstance(condition, Ge):
+                            greater = True
+                        k = Dummy('k')
+                        condition = Eq(condition.lhs, k) if isinstance(condition.lhs, RandomIndexedSymbol)\
+                            else Eq(condition.rhs, k)
+                        total = Sum(self.probability(condition, new_given_condition), (k, next_state + upper, self.state_space._sup))
+                        return Piecewise((total, rv[0].key > min_key_rv.key), (Probability(condition), True)) if greater\
+                            else Piecewise((1 - total, rv[0].key > min_key_rv.key), (Probability(condition), True))
+
+                elif isinstance(condition, Or):
+                    return sum([self.probability(expr, given_condition, evaluate, **kwargs)
+                            for expr in condition.args])
+                else:
+                    return Probability(condition, given_condition)
+
```

Comment:
The algorithm is the same, right?

Reply:
Why the original one isn't working for symbolic times or symbolic states?

Reply:
Right now, there is no way to calculate probability for `P(Eq(Y[a], b), Eq(Y[c], d))` because the current implementation uses `as_set` which throws an error for symbolic state spaces. This is the reason I added this code. Also, the algorithm is the same for queries like `P(Eq(Y[a], Y[b]), Eq(Y[c], d))` but `Sum` is throwing an error for some reason. We might have to investigate about this

Reply:
Can you define two functions(kind of private) and dispatch the call to one of these based on whether the query contains symbolic states or not? I mean just decouple them.

Reply:
Yeah sure, I'll do that

Reply:
I've added a function `symbolic_probability` to calculate probabilities for  symbolic expressions.
While doing this, I noticed a bug in the algorithm for calculating queries of the type `P(Gt(Y[a], Y[b]), Eq(Y[c], d))`.
``` python
>>> from sympy import *
>>> from sympy.stats import *
>>> T = Matrix([[Rational(5, 10), Rational(3, 10), Rational(2, 10)], [Rational(2, 10), Rational(7, 10), Rational(1, 10)], [Rational(3, 10), Rational(3, 10), Rational(4, 10)]])
>>> Y = DiscreteMarkovChain("Y", [0, 1, 2], T)
>>> P(Gt(Y[5], Y[7]), Eq(Y[2], 1))
6657/25000
>>> P(Gt(Y[7], Y[5]), Eq(Y[2], 1))
6657/25000 #Both can't be equal!
#Happens for all relational conditions except Eq which is symmetric
```
The bug is because of these lines - 
https://github.com/sympy/sympy/blob/e0ec854c452e9b83f77fd6577b0f1c84786eaf70/sympy/stats/stochastic_process_types.py#L551-L553

If the key of the `lhs` is less than that of `rhs`, we swap them, but we aren't flipping the inequality from > to < (or any other pair for that matter). I've fixed this as well. Please let me know if anything is not clear

Reply:
The build has timed out. Can you please look into it?